onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -1,69 +1,87 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include "core/common/common.h"
 #include "core/optimizer/constant_folding.h"
 #include "core/graph/graph_utils.h"
 #include "core/optimizer/optimizer_execution_frame.h"
 #include "core/framework/op_kernel.h"
 #include "core/framework/ml_value.h"
 
+using namespace ::onnxruntime::common;
+
 namespace onnxruntime {
 
-Status ConstantFolding::Apply(Graph& graph, Node& node, bool& modified, bool& deleted) {
-  // Create execution frame for executing constant nodes.
-  OptimizerExecutionFrame::Info info({&node}, graph.GetAllInitializedTensors());
+Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  auto& order = graph_viewer.GetNodesInTopologicalOrder();
 
-  std::vector<int> fetch_mlvalue_idxs;
-  for (const auto* node_out : node.OutputDefs()) {
-    fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
-  }
+  for (NodeIndex i : order) {
+    auto* node = graph.GetNode(i);
+    if (!node) {
+      return Status(ONNXRUNTIME, INVALID_ARGUMENT);
+    }
 
-  OptimizerExecutionFrame frame(info, fetch_mlvalue_idxs);
+    // TODO Further investigate if this is needed here.
+    ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level));
```

Comment:
Recurse is to apply transformer to sub graph. So, the issue which constant folding can't apply control flow can be workaround by removing the line of code above. If so, the code node->ImplicitInputDefs().size() != 0 below can be removed.

Reply:
The problem with remove Recurse is that constant folding will still be triggered on the control flow operators and will fail since it will require the OpKernelContextInternal.
Also, see my answer to Ashwini above -- there should not be a problem doing constant folding inside subgraphs, the problem are only the control flow operators.

Reply:
I know the problem is only the control flow operators. I was thinking probably now only control flow operators have sub graphs.

Reply:
The latest check I use is looking whether the current node has subgraphs (at the moment probably only control flow operators as you say).
But what I meant is that if you have an If node, constant folding will be applied and will fail today, because we will call Compute on it. Removing Recurse will not prevent that. Removing Recurse will only avoid doing constant folding on the If's subgraphs and we don't want to avoid that. Constant folding should be done and is supported on the subgraphs.

Reply:
Agree.